### PR TITLE
Update on install

### PR DESCRIPTION
**Pull request by kinnison**
This is a semi-prototype, but in theory ready-for-showtime attempt at resolving #2330

Opinions from @Manishearth and @mcclure are invited.

### CODE REVIEW

**Comment by Manishearth** [comment]
So I think _always_ prompting about this is a mistake, since all users will have to care about this.

We should either:

 - detect that there is an existing install at the beginning, and _only then_ prompt "There is an existing toolchain installed. Update it? Y/n"
 - run `rustup check` after installation and prompt if necessary

I'd prefer the former UX since then there's only one block of prompts.

**Comment by rbtcollins** [comment]
I've commented on https://github.com/rust-lang/rustup/issues/2330 - I don't think prompting is ever desirable here.

**Comment by kinnison** [comment]
@Manishearth How do you feel about this prototype?

**Review by Manishearth — APPROVED** [review]
If i'm reading this right this unconditionally does the equivalent of `rustup update <default>` on install unless otherwise specified?

Seems good to me!

**Comment by kinnison** [comment]
Yep that's the intent.  Okay, @rbtcollins how about you? Are you okay with this?

**Comment by mcclure** [comment]
I'm sorry, I don't know how to test this. I'm sure y'all know what's right for your project tho.

This said, I don't understand why you're trying so hard to avoid a prompt given when last I ran it, rustup.sh already shows a prompt?

**Comment by Manishearth** [comment]
@mcclure the initial prototype prompted for _every_ install, even if there wasn't an existing installation to update, which means a lot of people need to go through an unnecessary prompt.

The current prototype never prompts, which is also fine IMO, it's hard to see a use case for not updating an existing install when you are installing rustup (as opposed to updating it)

File: src/cli/setup_mode.rs [hunk 1]
```diff
@@ -81,6 +81,11 @@ pub fn main() -> Result<utils::ExitCode> {
                 .multiple(true)
                 .use_delimiter(true),
         )
+        .arg(
+            Arg::with_name("no-update-default-toolchain")
```
Comment (rbtcollins) [inline src/cli/setup_mode.rs:85]: I feel this is a bit verbose. We have --no-self-update; perhaps just --no-update ?

File: tests/cli-inst-interactive.rs [hunk 2]
```diff
@@ -14,6 +14,12 @@ use std::io::Write;
 use std::process::Stdio;
 use std::sync::Mutex;
 
+macro_rules! for_host {
```
Comment (rbtcollins) [inline tests/cli-inst-interactive.rs:17]: I don't really feel that a used-once macro that just forms a function once, used in a single test is really pulling its weight - but if you feel it is going to be used again I'm cool with it being added.

**Review by rbtcollins — APPROVED** [review]

File: src/cli/setup_mode.rs (hunk 1 above)
Comment (kinnison) [reply src/cli/setup_mode.rs:85]: I will have a ponder.  `--no-update` feels a little odd but the terseness may be okay.  I'll reconsider in the face of the rest of the args, it is a bit long yes.

File: tests/cli-inst-interactive.rs (hunk 2 above)
Comment (kinnison) [reply tests/cli-inst-interactive.rs:17]: I'm intending, once this is merged, to file an e-easy e-mentor help-wanted to move those macros into `mock/clitools.rs`

**Pull request merged**